packages/assembly/index.ts

```diff
@@ -850,7 +850,7 @@ export namespace ASON {
       // all the references have been allocated, let's get entry 0 and validate type info
       let entry0 = changetype<usize>(entryMap.get(0));
       if (isReference<T>()) {
-        assert(getObjectType(entry0) == idof<T>());
+        assert(getObjectType(entry0) == idof<T>() || __instanceof(entry0, idof<T>()));
```

Comment:
This is the line in question. `__instanceof` ships with the `assemblyscript` runtime as a means to check `a instanceof B` when `B` is a base class.